Add docs for order expiration (#1192)
saleor/saleor-docs: docs/developer/order-expiration.mdx

---
title: Order Expiration
---

:::info

expireOrdersAfter was introduced in **Saleor 3.13**

deleteExpiredOrdersAfter was introduced in **Saleor 3.14**

:::

:::caution

This feature is in the **Feature Preview** stage, which means that it is available for experimentation and
feedback. However, it is still undergoing development and is subject to modifications.

:::

## Introduction

> Saleor allows finalizing checkout without payment. This can be achieved by setting `orderSettings.allowUnpaidOrders` for a given `Channel` to `true` and calling `checkoutComplete` without any payment/transaction object or by using the `orderCreateFromCheckout` mutation with first order then payment flow.

Reviewer: ```suggestion
Saleor allows finalizing checkout without payment. This can be achieved in two ways:
- by setting `orderSettings.allowUnpaidOrders` for a given `Channel` to `true` and calling `checkoutComplete` without any payment/transaction object
- by using the `orderCreateFromCheckout` mutation with the first order and then payment flow.
```

    channel {
      id
      orderSettings {
        expireOrdersAfter
        deleteExpiredOrdersAfter
      }
    }
    errors {
      field
      code
      message
    }
  }
}
```

```json
{
  "data": {
    "channelUpdate": {
      "channel": {
        "id": "Q2hhbm5lbDoyMjQz",
        "orderSettings": {
          "expireOrdersAfter": 360,
          "deleteExpiredOrdersAfter": 30
        }
      },
      "errors": []
    }
  },
  "extensions": {
    "cost": {
      "requestedQueryCost": 0,
      "maximumAvailable": 50000
    }
  }
}
```

## How it works

Once an unpaid order is created (via [`checkoutComplete`](../../../api-reference/checkout/mutations/checkout-complete) or [`orderCreateFromCheckout`](api-reference/orders/mutations/order-create-from-checkout.mdx)), it remains in `UNCONFIRMED` status. If the customer does not complete the payment within `expireOrdersAfter` minutes after that, the status will change to `EXPIRED` and stock allocation will be released.
After `deleteExpiredOrdersAfter` days orders in `EXPIRED` status will be deleted and disappear from the order list.

## How to disable order expiration

To disable order expiration, set `expireOrdersAfter` to 0.

> When `expireOrdersAfter` is set to 0, the orders will not expire. Since there will be no orders with status `EXPIRED` Saleor will not automatically delete such orders.

Reviewer: ```suggestion
When `expireOrdersAfter` is set to 0, the orders will not expire. Since there will be no orders with the status `EXPIRED` Saleor will not automatically delete such orders.
```

---
title: Order Expiration
---

:::info

> expireOrdersAfter was introduced in **Saleor 3.13**

Reviewer 2: Camel case here, doesn't say anything here - I mean you don't link it to Channel type etc so it won't give you any hint. 
I would change it to a normal name like "Order expiration has been introduced in XX"

deleteExpiredOrdersAfter was introduced in **Saleor 3.14**

:::

:::caution

This feature is in the **Feature Preview** stage, which means that it is available for experimentation and
feedback. However, it is still undergoing development and is subject to modifications.

:::

## Introduction

Saleor allows finalizing checkout without payment. This can be achieved in two ways:

- by setting `orderSettings.allowUnpaidOrders` for a given `Channel` to `true` and calling `checkoutComplete` without any payment/transaction object

> - by using the `orderCreateFromCheckout` mutation with first order then payment flow.

Reviewer 2: I would add `protected` here - to clarify that creating an order via this mutation requires permission. 

`with first order then payment flow.` -> I would change it to something as you have in previous step: `without any payment/transaction object` (would be also good to use names from api-reference, (and links them with api-reference section): `Payment`/`TransactionItem` object)

> If the payment process is canceled by the customer and the order remains unpaid for some time, it is possible to set automatic stock releases and remove the order.

Reviewer 2: "and remove the order." - would be good to be more precise here. I mean we first mark the order as expired, then after `deleteExpiredOrdersAfter` we delete it. I think it is worth mentioning that this process is based on two steps: `expire` -> `delete`.

Reviewer 2: (I see that you clarify it more below, but still think it would be worth explicitly writing that the process contains two steps.)

## Configuration

To set the expiration of orders, specify the number of minutes after which the orders should expire by setting `expireOrdersAfter`.
Additionally, the `deleteExpiredOrdersAfter` value can be modified. By default, this is set to 60 days, but it can be changed to anywhere between 1 and 120 days.

```graphql
mutation ChannelUpdate {
  channelUpdate(
    id: "Q2hhbm5lbDoyMjQz"
    input: {
      orderSettings: { expireOrdersAfter: 360, deleteExpiredOrdersAfter: 30 }
    }
  ) {
    channel {
      id
      orderSettings {
        expireOrdersAfter
        deleteExpiredOrdersAfter
      }
    }
    errors {
      field
      code
      message
    }
  }
}
```

```json
{
  "data": {
    "channelUpdate": {
      "channel": {
        "id": "Q2hhbm5lbDoyMjQz",
        "orderSettings": {
          "expireOrdersAfter": 360,
          "deleteExpiredOrdersAfter": 30
        }
      },
      "errors": []
    }
  },
  "extensions": {
    "cost": {
      "requestedQueryCost": 0,
      "maximumAvailable": 50000
    }
  }
}

> ```

Reviewer 2: You could add here a sentence or two that will explain that as a result of this example, your unpaid orders will expire after 360 minutes, and will be deleted after 30 days.

---
title: Order Expiration
---

:::info

expireOrdersAfter was introduced in **Saleor 3.13**

deleteExpiredOrdersAfter was introduced in **Saleor 3.14**

:::

:::caution

This feature is in the **Feature Preview** stage, which means that it is available for experimentation and
feedback. However, it is still undergoing development and is subject to modifications.

:::

## Introduction

> Saleor allows to finalize transactions without payment (order first, then payment). If the customer cancel payment process and order remains unpaid for some time, there is possibility to set automatic stock releases and remove order.

Reviewer 2: IMHo `finalize transactions` doesn't say to me, that we are talking about checkout->order process. 

> If the customer cancel payment process and order remains unpaid for some time, there is possibility to set automatic stock releases and remove order.

This only one use case. There is also `allowUnpaidOrders` that would allow you to convert checkout into order without any payment by calling `checkoutComplete`. So you will have unpaid orders, that will wait for a payment or expiration

## Configuration

> To set the expiration of orders, turn on the `automaticallyConfirmAllNewOrders` option in the channel settings and specify the number of minutes after which the orders should expire by setting `expireOrdersAfter`.

Reviewer 2: How `automaticallyConfirmAllNewOrders`is related to this process? 🤔

Additionally, you can modify `deleteExpiredOrdersAfter`. By default, this is set to 60 days, but you can change this value to anywhere between 1 and 120 days.

```graphql
mutation ChannelUpdate {
  channelUpdate(
    id: "Q2hhbm5lbDoyMjQz"
    input: {
      orderSettings: {
        expireOrdersAfter: 360
        deleteExpiredOrdersAfter: 30
        automaticallyConfirmAllNewOrders: true
      }
    }
  ) {
    channel {
      id
      orderSettings {
        expireOrdersAfter
        deleteExpiredOrdersAfter
        automaticallyConfirmAllNewOrders
      }
    }
    errors {
      field
      code
      message
    }
  }
}
```

```json
{
  "data": {
    "channelUpdate": {
      "channel": {
        "id": "Q2hhbm5lbDoyMjQz",
        "orderSettings": {
          "expireOrdersAfter": 360,
          "deleteExpiredOrdersAfter": 30,
          "automaticallyConfirmAllNewOrders": true
        }
      },
      "errors": []
    }
  },
  "extensions": {
    "cost": {
      "requestedQueryCost": 0,
      "maximumAvailable": 50000
    }
  }
}
```

## How it works

> Once an unpaid order is created (via [`checkoutCreate`](api-reference/checkout/mutations/checkout-create.mdx) or [`orderCreateFromCheckout`](api-reference/orders/mutations/order-create-from-checkout.mdx)), it remains in `UNCONFIRMED` status. If the customer does not complete the payment within `expireOrdersAfter` minutes after that, the status will change to `EXPIRED` and stocks allocation will be released.

Reviewer 2: > via [`checkoutCreate`]

It doesn't create an unpaid order, but new checkout instance. Is it correct?

After `deleteExpiredOrdersAfter` days orders in `EXPIRED` status will be deleted and disappear from the order list.

## How to disable order expiration

To disable order expiration, set `expireOrdersAfter` to 0.

> When `expireOrdersAfter` is set to 0, the order will not expire and will not be deleted.

Reviewer 2: > and will not be deleted.

I think this could be misleading, you could provide here more context, like explaining, that when `expireOrdersAfter` is turned off, the logic responsible for deleting expired orders won't make anything.

---
title: Order Expiration
---

:::info

Order expiration was introduced in **Saleor 3.13**

The removal of expired orders was introduced in **Saleor 3.14**

:::

:::caution

This feature is in the **Feature Preview** stage, which means that it is available for experimentation and
feedback. However, it is still undergoing development and is subject to modifications.

:::

## Introduction

Saleor allows finalizing checkout without payment. This can be achieved in two ways:

- by setting [`orderSettings.allowUnpaidOrders`](api-reference/miscellaneous/objects/order-settings.mdx#ordersettingsallowunpaidordersboolean--) for a given `Channel` to `true` and calling [`checkoutComplete`](api-reference/checkout/mutations/checkout-complete.mdx) without any [`Payment`](api-reference/payments/objects/payment.mdx)/[`TransactionItem`](api-reference/payments/objects/transaction-item.mdx) object
- by calling the protected mutation [`orderCreateFromCheckout`](api-reference/orders/mutations/order-create-from-checkout.mdx) without any [`Payment`](api-reference/payments/objects/payment.mdx)/[`TransactionItem`](api-reference/payments/objects/transaction-item.mdx) object.

> If the customer cancels the payment process and the order remains unpaid for some time, it is possible to set automatic stock releases and remove the order. Saleor will first mark orders as expired and later will remove those orders.

Reviewer 2: ```suggestion
If the customer cancels the payment process or the order remains unpaid for some time, it is possible to set automatic stock releases and remove the order. Saleor will first mark orders as expired and later will remove those orders.
```
OR
```suggestion
If the order remains unpaid for some time, it is possible to set automatic stock releases and remove the order. Saleor will first mark orders as expired and later will remove those orders.
```
